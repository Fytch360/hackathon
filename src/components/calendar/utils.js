import {
    startOfWeek,
    endOfWeek,
    startOfMonth,
    endOfMonth,
    startOfYear,
    endOfYear,
    eachDayOfInterval,
    eachWeekOfInterval,
    eachMonthOfInterval,
    isSameMonth,
    startOfDecade,
    addYears,
    setMinutes,
    setHours,
    parseISO,
    formatISO,
    parse,
    isWithinInterval,
    format as formatDate,
    isBefore,
    isSameDay
} from 'date-fns'
import { chunk } from 'lodash'

import {
    locale,
    MIN_YEAR_RESTRICTION_DATE,
    MAX_YEAR_RESTRICTION_DATE,
    MONTHS_IN_ROW,
    YEARS_IN_ROW,
    YEARS_IN_DECADE,
    DATE_DEFAULT_HOUR,
    DATE_DEFAULT_MINUTE,
    commonMaskOptions,
    MaskFormat,
    MaskPattern,
    InputMode
} from './constants'

const options = {
    locale,
}

export const setDefaultHHMM = (date) =>
    setMinutes(setHours(date, DATE_DEFAULT_HOUR), DATE_DEFAULT_MINUTE)
export const capitalize = (str) => str.charAt(0).toUpperCase() + str.slice(1)

// Возвращает массив массивов из дней для указанной даты
export const getMonthDates = (date, viewModeLight) => {
    const dateParsed = parseISO(date)
    const monthDates = []

    const weeks = eachWeekOfInterval(
        {
            start: startOfMonth(dateParsed),
            end: endOfMonth(dateParsed),
        },
        options
    )

    weeks.forEach((week, weekIndex) => {
        monthDates[weekIndex] = []

        const days = eachDayOfInterval({
            start: startOfWeek(week, options),
            end: endOfWeek(week, options),
        })

        days.forEach((day, dayIndex) => {
            monthDates[weekIndex][dayIndex] = isSameMonth(dateParsed, day)
                ? formatISO(setDefaultHHMM(day))
                : null
        })
    })

    return monthDates
}

// Возвращает массив массивов из месяцев для указанной даты
export const getYearDates = (date) => {
    const dateParsed = parseISO(date)
    const start = startOfYear(dateParsed)
    const end = endOfYear(dateParsed)
    const yearDates = eachMonthOfInterval({ start, end }).map((month) =>
        formatISO(setDefaultHHMM(month))
    )

    return chunk(yearDates, MONTHS_IN_ROW)
}

// Возвращает массив массивов из десятилетия для указанной даты
export const getDecadeDates = (date) => {
    const dateParsed = parseISO(date)
    const decadeDates = new Array(YEARS_IN_DECADE).fill().map((_, index) => {
        const startDateOfDecade = startOfDecade(dateParsed)
        const nextDateOfDecade = addYears(
            startDateOfDecade,
            parseInt(index, 10)
        )

        return formatISO(setDefaultHHMM(nextDateOfDecade))
    })

    return chunk(decadeDates, YEARS_IN_ROW)
}

export const createMaskOptions = (mode, restriction, mask) => ({
    ...commonMaskOptions,
    pattern: mask || MaskPattern[mode],
    min: Array.isArray(restriction) || !restriction.start ? commonMaskOptions.min : restriction.start,
    max: Array.isArray(restriction) || !restriction.end ? commonMaskOptions.max : restriction.end,
    format: (date) => formatDate(parseISO(date), MaskFormat[mode], { locale }),
    parse: (string) => {
        if (mode === InputMode.DEFAULT) {
            const [day, month, year] = string.split('.')
            return new Date(year, month - 1, day)
        }

        return parse(mode === InputMode.DAY ? string.match(/\d+/g).pop() : string, MaskFormat[mode], new Date(), { locale })
    }
})

export const getInitialError = (date, restriction, validation) => {
    if (!date) {
        return ''
    }

    const dateParsed = parseISO(date)
    const error = Array.isArray(restriction) ? !restriction.some((available) => validation(available, dateParsed)) : !isWithinInterval(dateParsed, restriction)
    
    return error ? ' ' : ''
}

export const getInitialRestriction = (restriction) => {
    if (Array.isArray(restriction) && restriction.length) {
        return restriction
    }

    return {
        start: restriction.start || MIN_YEAR_RESTRICTION_DATE,
        end: restriction.end || MAX_YEAR_RESTRICTION_DATE
    }
}

export const getBackgroundColor = ({ theme, isStartDate, isEndDate, disabled }) => {
    if ((isStartDate || isEndDate) && !disabled) {
        return 'transparent'
    }

    if (disabled) {
        return theme.calendarButtonInvalid
    }

    return theme.calendarButtonWithin
}

export const swapDates = (state) => {
    const { start, end } = state

    // Если дата конца раньше чем дата начала,
    // то свапаем даты местами
    if (isBefore(parseISO(end), parseISO(start))) {
        state.start = end
        state.end = start
    }
}
