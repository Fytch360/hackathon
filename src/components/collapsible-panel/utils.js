import React from 'react'

const passProps = (children, props) =>
    typeof children.type === 'function' ? React.cloneElement(children, { ...props }) : children

/**
 * @param {function, node} children
 * @param {object} props, что будут прокинуты внутрь children, если он является ReactNode(function), иначе вернет как есть
 * @returns {function, node} всё то же самое, только с прокинутыми props (по условию выше)
 */

export const passPropsToChildren = (children, props) =>
    children && children.map
        ? children.map((item, index) => passProps(item, { key: index, ...props }))
        : passProps(children, { ...props })
