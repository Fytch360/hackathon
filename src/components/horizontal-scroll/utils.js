export const easeInOutQuad = (time, start, change, duration) => {
    // comment: лучше не скомпоновалось
    /* eslint-disable no-mixed-operators */
    const half = 2
    let halvedTime = time / (duration / half)
    if (halvedTime < 1) {
        return (change / half) * halvedTime * halvedTime + start
    }
    halvedTime -= 1
    return -change / half * (halvedTime * (halvedTime - half) - 1) + start
}

export const smoothScroll = (element, change, duration, cb) => {
    const start = element.scrollLeft
    const el = element
    let currentTime = 0
    const increment = 20

    const animateScroll = () => {
        currentTime += increment
        el.scrollLeft = easeInOutQuad(currentTime, start, change, duration)
        if (currentTime < duration) {
            requestAnimationFrame(animateScroll)
        } else {
            cb()
        }
    }
    animateScroll()
}

export const leftCheck = (scroll) => {
    if (!scroll) {
        return false
    }

    return scroll.scrollLeft > 0
}

export const rightCheck = (scroll) => {
    if (!scroll) {
        return false
    }

    // В IE scrollWidth больше clientWidth на 1px, даже если нет скролла, поэтому  > 1, а не > 0.
    return Math.floor(scroll.scrollWidth - scroll.clientWidth - scroll.scrollLeft) > 1
}

export const scrollToChild = (scroll, target, duration, cb) => {
    const scrollViewPortSize = scroll.clientWidth
    const targetSize = target.scrollWidth
    const scrollDistance = scroll.scrollLeft
    const targetScrollOffset = target.offsetLeft
    
    const scrollEdge = (scrollDistance < targetScrollOffset) && (targetSize < scrollViewPortSize) ? 'right' : 'left'

    const targetOffset = target.getBoundingClientRect()?.[scrollEdge]
    const scrollOffset = scroll.getBoundingClientRect()?.[scrollEdge]

    const distanceToTarget = targetOffset - scrollOffset

    smoothScroll(scroll, distanceToTarget, duration, cb)
}
