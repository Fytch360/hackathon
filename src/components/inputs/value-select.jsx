import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { useField } from 'formik'
import { useTranslation } from 'react-i18next'
import _ from 'lodash'
import {
    LabeledValueSelect,
    ValueOption,
    ValueSubheader
} from './value-select'
import { Caption } from './typography'

import ViewLabel from './view-label'

const ValueSelect = ({ name, options, onChange, viewOnly, type, requisites, ...props }) => {
    const [field, meta, helpers] = useField({ name })
    const totalPaymentAmount = useField({ name: 'totalPaymentAmount' })
    const commissionField = useField({ name: 'commission' })
    const { t } = useTranslation()

    const actionRoute = _.get(requisites?.data, 'action.route', '')
    const isAutoPayment = ['AUTOPAY_CREATE_SAVE', 'AUTOPAY_EDIT_SAVE'].includes(actionRoute)
    const selectedSource =
        field.name === 'source' && options.find((option) => option.value === field.value)
    const totalAmountValueWithCommission = commissionField[0].value
        ? parseFloat(totalPaymentAmount[0].value) + parseFloat(commissionField[0].value)
        : parseFloat(totalPaymentAmount[0].value)

    const handleChange = useCallback(
        (value) => {
            helpers.setTouched(value)
            helpers.setValue(value)
            onChange(value, _.find(options, { value }))
        },
        [helpers, onChange, options]
    )

    if (viewOnly) {
        const selectedOption = _.find(options, ['value', field.value])
        const displayValue =
            type === 'list'
                ? selectedOption?.title
                : `${selectedOption?.title} ${selectedOption?.description}`

        return (
            <ViewLabel label={props.label} fontWeight="semibold">
                {displayValue || field.value || '-'}
            </ViewLabel>
        )
    }

    return (
        <>
            <LabeledValueSelect
                error={meta.touched ? meta.error : ''}
                translations={{ placeholder: t('common:validation.select') }}
                {...field}
                {...props}
                onChange={handleChange}
            >
                {options.map(({ value, title, isSubheader, ...rest }) =>
                    isSubheader ? (
                        <ValueSubheader key={title} title={t(title)} />
                    ) : (
                        <ValueOption key={value} value={value} title={title} {...rest} />
                    )
                )}
            </LabeledValueSelect>
            {totalAmountValueWithCommission > selectedSource?.amount && !isAutoPayment && (
                <Caption indent="zero" colorScheme="warning">
                    {t('payments:payments.notEnough')}
                </Caption>
            )}
        </>
    )
}

ValueSelect.propTypes = {
    viewOnly: PropTypes.bool,
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    requisites: PropTypes.array.isRequired,
    options: PropTypes.array,
    onChange: PropTypes.func,
    type: PropTypes.string
}

ValueSelect.defaultProps = {
    viewOnly: false,
    options: [],
    onChange: () => null,
    type: null
}

export default ValueSelect
