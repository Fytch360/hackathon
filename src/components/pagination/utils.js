import { useState } from 'react'

const PAGE_2 = 2
const PAGE_4 = 4
const PAGE_5 = 5

export const getMediaListeners = (media, cb) => {
    if (media.addEventListener) {
        return {
            addMediaListener: () => media.addEventListener('change', cb),
            removeMediaListener: () => media.removeEventListener('change', cb)
        }
    }

    // https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/addListener
    // backwards compatibility IE, Opera
    return {
        addMediaListener: () => media.addListener(cb),
        removeMediaListener: () => media.removeListener(cb)
    }
}

export const usePagination = ({
    count = 1,
    defaultPage = 1,
    page,
    onChange: handleChange,
    mobile = false
}) => {
    const [pageState, setPageState] = useState(page || defaultPage)

    const handleClick = (event, value) => {
        setPageState(value)
        if (handleChange) {
            handleChange(value)
        }
    }

    const start = Math.max(
        Math.min(
            // Если по центру нужно с -1 элемента начать
            pageState > PAGE_5 ? pageState - 1 : 1,
            // Первая цифра за троеточием в конце
            count - PAGE_4,
        ),
        // Первая цифра перед троеточием в начале
        PAGE_2,
        // Если в конце списка, то значение count - 1
        pageState > count - PAGE_5 ? count - PAGE_4 : 1,
    )

    const end = Math.min(
        Math.max(
            // Если по центру, то +1 элемент рисуем
            pageState > PAGE_5 ? pageState + 1 : 1,
            // Если в конце списка, то значение count - 1
            pageState > count - PAGE_5 ? count - 1 : 1,
            // Последняя цифра за троеточием в начале
            PAGE_5,
        ),
        // Последняя цифра после троеточия в конце
        count - 1,
    )

    // Создаем массив с числами от start до end
    // https://dev.to/namirsab/comment/2050
    const range = () => {
        const length = end - start + 1
        return length >= 0 && Array.from({ length }, (_, i) => start + i)
    }

    const mobileRange = () => {
        const length = count < 3 ? count : 3
        const page = Math.max(
            count === pageState ? pageState - 2 : pageState - 1,
            1,
        )
        return length >= 0 && Array.from({ length }, (_, i) => page + i)
    }

    const rangeArr = range()
    const mobileRangeArr = mobileRange()

    const itemList = mobile ? [
        count > 1 ? 'first' : [],
        // Кнопка назад
        count > 1 ? 'previous' : [],
        ...mobileRangeArr,
        count > 1 ? 'next' : [],
    ] :
        [
        // Кнопка назад
            count > 1 ? 'previous' : [],
            // 1 страница
            1,
            // Вставляем многоточие, если нужная страница не первые пять
            ...(start >= PAGE_5 ? ['doublePrevious'] : []),
            // Вставляем середину страницы которую должно быть видно
            ...(rangeArr || []),
            // Вставляем многоточие, если нужная страница не последние пять
            ...(end <= count - PAGE_4 ? ['doubleNext'] : []),
            // Последняя страница
            count > 1 ? count : [],
            count > 1 ? 'next' : [],
        ]

    const buttonPage = (type) => {
        switch (type) {
            case 'doublePrevious':
                return pageState - 3
            case 'previous':
                return pageState - 1
            case 'next':
                return pageState + 1
            case 'doubleNext':
                return pageState + 3
            case 'first':
                return 1
            default:
                return null
        }
    }

    const items = itemList.map((item) => typeof item === 'number'
        ? {
            onClick: (event) => {
                handleClick(event, item)
            },
            type: 'page',
            pageItem: item,
            selected: item === pageState,
            'aria-current': item === pageState ? 'true' : 'false',
        }
        : {
            onClick: (event) => {
                handleClick(event, buttonPage(item))
            },
            type: item,
            pageItem: buttonPage(item),
            selected: false,
            disabled:
            (item.indexOf('previous') === -1 || item.indexOf('next') === -1 || item.indexOf('first') === -1) &&
              (item === 'next' ? pageState >= count : pageState <= 1),
        })


    return { items }
}
