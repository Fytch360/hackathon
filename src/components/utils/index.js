export { formatNumber } from './format-number'
export { makeMoneyMask } from './make-money-mask'
export { formatValueByMask } from './format-value'
export { formatLocalPhone } from './format-local-phone'
export { propStyles, variantStyles } from './apply-styles'
export { scrollToElement } from './scroll-to'
export { getFilteredComponent } from './get-filtered-component'
export { getDisplayName } from './get-display-name'
