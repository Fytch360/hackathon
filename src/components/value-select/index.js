export { ValueSelect, LabeledValueSelect } from './value-select'
export { ValueOption } from './value-option'
export { ValueSubheader } from './value-subheader'
