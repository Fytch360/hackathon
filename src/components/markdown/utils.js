// comment: кейсинг предоставляет Remarkable
/* eslint-disable camelcase */
import _ from 'lodash'
import Remarkable from 'remarkable'
import { escapeHtml, replaceEntities } from 'remarkable/lib/common/utils'

import { externalRegular as ExternalIcon } from '../icon/common'

import { textColorParcer, textColorRule } from './text-color'

const options = {
    html: false,
    xhtmlOut: false,
    breaks: false,
    langPrefix: 'language-',
    linkify: false,
    typographer: true,
    quotes: '«»‘’'
}

const internalTestRegExp = /^[#./]/
const lastWorldRegExp = /\S*$/
const nbspRegExp = /\u00A0/g
const nbspHtmlEntity = '&nbsp;'

export const isExternalUrl = (url = '') =>
    !!url && !internalTestRegExp.test(url)

export const linkOpen = (tokens, idx) => {
    const href = escapeHtml(tokens[idx].href)
    const isExternal = isExternalUrl(href)
    const title = tokens[idx].title
        ? ` title="${escapeHtml(replaceEntities(tokens[idx].title))}"`
        : ''

    if (!isExternal) {
        return `<a href="${href}"${title}><span>`
    }

    const externals = ' target="_blank" rel="noopener noreferrer"'
    const content = escapeHtml(tokens[idx + 1].content)
    const droppedLast = content.replace(lastWorldRegExp, '')

    return `<a href="${href}"${title}${externals}>${droppedLast ? '<span>' : ''}`
}

export const linkClose = (tokens, idx) => {
    // comment: приходят все токены, а href лежит в два токена слева: <a> и текст
    const href = escapeHtml(tokens[idx - 2].href)
    if (!isExternalUrl(href)) {
        return '</span></a>'
    }

    const content = escapeHtml(tokens[idx - 1].content)
    const droppedLast = content.replace(lastWorldRegExp, '')
    const lastWord = content.match(lastWorldRegExp)[0]

    return `${droppedLast ? '</span>' : ''}<span>${lastWord}${ExternalIcon}</span></a>`
}

export const text = (tokens, idx) => {
    const content = escapeHtml(tokens[idx].content)
    const href = escapeHtml(tokens[idx - 1]?.href)

    if (!isExternalUrl(href)) {
        return content
    }

    return content.replace(lastWorldRegExp, '')
}

export const replaceNbsp = (stateCore) => {
    stateCore.src = stateCore.src?.replace(nbspRegExp, nbspHtmlEntity)
}

const inlineDisable = [
    'backticks',
    'ins',
    'mark',
    'footnote_inline',
    'footnote_ref'
]
const blockDisable = ['blockquote', 'code', 'fences']
const blockDisableAdditional = ['deflist', 'table']

const configMdCommon = () => {
    const markdown = new Remarkable('full', options)

    markdown.core.ruler.before('block', 'nbspReplacer', replaceNbsp)
    markdown.inline.ruler.push('text_color', textColorRule)
    markdown.inline.ruler.disable(inlineDisable)
    markdown.renderer.rules.link_open = linkOpen
    markdown.renderer.rules.link_close = linkClose
    markdown.renderer.rules.text = text
    markdown.renderer.rules.color = textColorParcer

    return markdown
}

const configMdFull = () => {
    const markdownFull = configMdCommon()
    markdownFull.block.ruler.disable(blockDisable)

    return markdownFull
}

const configMdShort = () => {
    const markdownShort = configMdCommon()
    // comment: нужно поменять,но убедится что ничего не сломалось
    markdownShort.block.ruler.disable(blockDisable.concat(blockDisableAdditional))

    // ignores images completely
    markdownShort.renderer.rules.image = _.stubString

    return markdownShort
}

export const markdownFull = configMdFull()
export const markdownShort = configMdShort()
