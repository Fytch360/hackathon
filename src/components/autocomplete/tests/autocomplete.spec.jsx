import React from 'react'
import { mount } from 'enzyme'

import { Autocomplete } from '..'
import { ValueOption } from '../../value-select'
import { keyCodes } from '../../utils/dropdown-utils'

describe('<Autocomplete />', () => {
    const optionsExample = [
        { title: 'Title 0', value: 'Value 0' },
        { title: 'Title 1', value: 'Value 1' },
        { title: 'Title 2', value: 'Value 2' },
        { title: 'Title 3', value: 'Value 3' },
        { title: 'Title 4', value: 'Value 4' }
    ]
    const propsExample = {
        id: 'example-id',
        placeholder: 'example-placeholder',
        size: 'lg',
    }
    let wrapper
    let onChangeMock

    beforeEach(() => {
        onChangeMock = jest.fn()

        wrapper = mount(
            <Autocomplete {...propsExample} onChange={onChangeMock}>
                {optionsExample.map(({ title, value }) =>
                    <ValueOption key={value} value={value} title={title} />
                )}
            </Autocomplete>
        )
    })

    test('Should be available', () => {
        expect(Autocomplete).toBeDefined()
    })

    test('Snapshot', () => {
        wrapper.setProps({ mode: 'loading' })

        expect(wrapper).toMatchSnapshot()
    })

    test('Should\'nt render Contents by default', () => {
        expect(wrapper.find('ContentsStyled').exists()).toBeFalsy()
    })

    test('Should render Contents when textField was focussed', () => {
        wrapper.find('input')
            .simulate('focus', { })

        expect(wrapper.find('ContentsStyled')).toHaveLength(1)
    })

    test('Should render noMatches translation in Contents if gets "noMatches" mode', () => {
        const noMatchesTranslationExample = 'blah-blah-blah-noMatches'

        wrapper.setProps({
            mode: 'noMatches',
            translations: { noMatches: noMatchesTranslationExample }
        })

        wrapper.find('input')
            .simulate('focus', { })

        expect(wrapper.find('ContentsStyled').text()).toBe(noMatchesTranslationExample)
    })

    test('Should render loadingError translation in Contents if gets "error" mode', () => {
        const loadingErrorTranslationExample = 'blah-blah-blah-loadingError'

        wrapper.setProps({
            mode: 'error',
            translations: { loadingError: loadingErrorTranslationExample }
        })

        wrapper.find('input')
            .simulate('focus', { })

        expect(wrapper.find('ContentsStyled').text()).toBe(loadingErrorTranslationExample)
    })

    test('Should render ContainedLoader in Contents if gets "loading" mode', () => {
        wrapper.setProps({ mode: 'loading' })

        wrapper.find('input')
            .simulate('focus', { })

        expect(wrapper.find('ContentsStyled').find('ContainedLoader')).toHaveLength(1)
    })

    test('Should return empty value if ESC button was pressed', () => {
        wrapper.find('input')
            .simulate('keydown', { keyCode: keyCodes.KEY_ESCAPE })
        
        expect(onChangeMock.mock.calls[0][0]).toBe('')
    })

    test('Should\'nt call onChange if pressed only enter key', () => {
        wrapper.find('input')
            .simulate('keydown', { keyCode: keyCodes.KEY_ENTER })

        expect(onChangeMock.mock.calls).toHaveLength(0)
    })

    test('Should return first option\'s value and title if pressed home key and then option was choosed', () => {
        wrapper.find('input')
            .simulate('keydown', { keyCode: keyCodes.KEY_HOME })
            .simulate('keydown', { keyCode: keyCodes.KEY_ENTER })
        
        expect(onChangeMock.mock.calls[0][0]).toBe(optionsExample[0].title)
        expect(onChangeMock.mock.calls[0][1]).toBe(optionsExample[0].value)
    })

    test('Should return last option\'s value and title if pressed end key and then option was choosed', () => {
        const lastOptionIndex = optionsExample.length - 1

        wrapper.find('input')
            .simulate('keydown', { keyCode: keyCodes.KEY_END })
            .simulate('keydown', { keyCode: keyCodes.KEY_ENTER })
        
        expect(onChangeMock.mock.calls[0][0]).toBe(optionsExample[lastOptionIndex].title)
        expect(onChangeMock.mock.calls[0][1]).toBe(optionsExample[lastOptionIndex].value)
    })

    test('Should return first option\'s value and title if pressed arrow down and then option was choosed', () => {
        wrapper.find('input')
            .simulate('keydown', { keyCode: keyCodes.KEY_ARROW_DOWN })
            .simulate('keydown', { keyCode: keyCodes.KEY_ENTER })
        
        expect(onChangeMock.mock.calls[0][0]).toBe(optionsExample[0].title)
        expect(onChangeMock.mock.calls[0][1]).toBe(optionsExample[0].value)
    })

    test('Should return penultimate option\'s value and title if pressed end key then pressed arrow up and then option was choosed', () => {
        const lastOptionIndex = optionsExample.length - 1

        wrapper.find('input')
            .simulate('keydown', { keyCode: keyCodes.KEY_END })
            .simulate('keydown', { keyCode: keyCodes.KEY_ARROW_UP })
            .simulate('keydown', { keyCode: keyCodes.KEY_ENTER })
        
        expect(onChangeMock.mock.calls[0][0]).toBe(optionsExample[lastOptionIndex - 1].title)
        expect(onChangeMock.mock.calls[0][1]).toBe(optionsExample[lastOptionIndex - 1].value)
    })

    test('Should\'nt render Contents if Tab key was pressed', () => {
        wrapper.find('input')
            .simulate('focus', {})
            .simulate('keydown', { keyCode: keyCodes.KEY_TAB })
        
        expect(wrapper.find('ContentsStyled').exists()).toBeFalsy()
    })
})
