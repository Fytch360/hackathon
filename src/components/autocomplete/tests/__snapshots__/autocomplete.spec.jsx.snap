// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete /> Snapshot 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-4 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  border: solid 1px;
  border-radius: 4px;
  -webkit-transition: 0.17s ease box-shadow;
  transition: 0.17s ease box-shadow;
  height: 66px;
  padding: 0;
  padding-right: 20px;
  padding-left: 20px;
}

.emotion-4:hover {
  border: solid 1px;
}

.emotion-6 {
  -webkit-tap-highlight-color: transparent;
  border-radius: 4px;
  width: 100%;
  outline: none;
  display: block;
  font-weight: 500;
  background-color: transparent;
  -webkit-appearance: textfield;
  -moz-appearance: textfield;
  -ms-appearance: textfield;
  appearance: textfield;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  margin: 0;
  margin-top: 0px;
  margin-bottom: 0px;
  margin-right: 0px;
  font-size: 19px;
  line-height: 24px;
}

.emotion-6::-ms-clear {
  display: none;
}

.emotion-6::-webkit-inner-spin-button,
.emotion-6::-webkit-outer-spin-button,
.emotion-6::-webkit-clear-button,
.emotion-6::-webkit-inner-spin-button,
.emotion-6::-webkit-calendar-picker-indicator {
  display: none;
}

.emotion-6:-webkit-autofill,
.emotion-6:-webkit-autofill:hover,
.emotion-6:-webkit-autofill:focus,
.emotion-6:-webkit-autofill:active {
  -webkit-text-fill-color: inherit;
  box-shadow: 0 0 0 1000px transparent inset;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
}

<Autocomplete
  id="example-id"
  mode="loading"
  onChange={[MockFunction]}
  placeholder="example-placeholder"
  size="lg"
>
  <OnClickOutside(Perimeter)
    disableOnClickOutside={true}
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    onClickOutside={[Function]}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
  >
    <Perimeter
      disableOnClickOutside={[Function]}
      enableOnClickOutside={[Function]}
      eventTypes={
        Array [
          "mousedown",
          "touchstart",
        ]
      }
      onClickOutside={[Function]}
      outsideClickIgnoreClass="ignore-react-onclickoutside"
      preventDefault={false}
      stopPropagation={false}
    >
      <WrapperStyled
        id="example-id-wrapper"
      >
        <div
          className="emotion-0 emotion-1"
          id="example-id-wrapper"
        >
          <TargetStyled
            aria-expanded={false}
            aria-haspopup="listbox"
            aria-owns="example-id-contents"
            role="combobox"
          >
            <div
              aria-expanded={false}
              aria-haspopup="listbox"
              aria-owns="example-id-contents"
              className="emotion-2 emotion-3"
              role="combobox"
            >
              <OmittedComponent
                aria-autocomplete="list"
                autoComplete="off"
                id="example-id"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="example-placeholder"
                size="lg"
                verticalMargin="zero"
              >
                <Component
                  aria-autocomplete="list"
                  autoComplete="off"
                  id="example-id"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="example-placeholder"
                  size="lg"
                  verticalMargin="zero"
                >
                  <WrapperStyled
                    focused={false}
                    horizontalPadding="inner"
                    size="lg"
                    verticalPadding="inner"
                  >
                    <span
                      className="emotion-4 emotion-5"
                      size="lg"
                    >
                      <TextFieldStyled
                        aria-autocomplete="list"
                        autoComplete="off"
                        horizontalMargin="zero"
                        horizontalMarginDirection="right"
                        horizontalPadding="inner"
                        id="example-id"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        placeholder="example-placeholder"
                        size="lg"
                        verticalMargin="zero"
                        verticalPadding="inner"
                      >
                        <input
                          aria-autocomplete="list"
                          autoComplete="off"
                          className="emotion-6 emotion-7"
                          id="example-id"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onClick={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          placeholder="example-placeholder"
                          size="lg"
                        />
                      </TextFieldStyled>
                    </span>
                  </WrapperStyled>
                </Component>
              </OmittedComponent>
            </div>
          </TargetStyled>
        </div>
      </WrapperStyled>
    </Perimeter>
  </OnClickOutside(Perimeter)>
</Autocomplete>
`;
